### PR TITLE
Add gRPC reverse proxy server example

### LINKED ISSUE #2353: Add gRPC reverse proxy server example
.. which is similar to our HTTP reverse proxy server example.

### PR DESCRIPTION
**Pull request by eottabom**
Motivation:

Add gRPC reverse proxy server example

Modifications:

Add gRPC reverse proxy server example with testcontainers.

Result:

- Closes #2353 .

### CODE REVIEW

**Comment by trustin** [comment]
Could you check the build failures and make sure the tests are not flaky?

File: examples/grpc/src/test/java/example/armeria/grpc/GrpcReverseProxyServerTest.java [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+
+    @Container
+    private static GenericContainer<?> envoy = new GenericContainer<>("envoyproxy/envoy:v1.22.0")
+            .withExposedPorts(9090)
+            .withFileSystemBind("./envoy/envoy.yaml", "/etc/envoy/envoy.yaml", BindMode.READ_WRITE);
+
+    @BeforeAll
+    static void startServer() {
+        server = Server.builder()
+                .http(8080)
```
Comment (trustin) [inline examples/grpc/src/test/java/example/armeria/grpc/GrpcReverseProxyServerTest.java]: Please use an ephemeral port instead, to avoid port collision.

File: examples/grpc/build.gradle [hunk 2]
```diff
@@ -16,6 +16,7 @@ dependencies {
     testImplementation libs.awaitility
     testImplementation libs.junit5.jupiter.api
     testImplementation libs.protobuf.java.util
+    testImplementation libs.testcontainers.junit.jupiter
```
Comment (trustin) [inline examples/grpc/build.gradle:19]: Please create a separate project for the example you're writing. Don't make an example do more than one thing.

File: examples/grpc/src/test/java/example/armeria/grpc/GrpcReverseProxyServerTest.java [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+
+    @Container
+    private static GenericContainer<?> envoy = new GenericContainer<>("envoyproxy/envoy:v1.22.0")
+            .withExposedPorts(9090)
+            .withFileSystemBind("./envoy/envoy.yaml", "/etc/envoy/envoy.yaml", BindMode.READ_WRITE);
+
+    @BeforeAll
+    static void startServer() {
+        server = Server.builder()
+                .http(18080)
```
Comment (trustin) [inline examples/grpc/src/test/java/example/armeria/grpc/GrpcReverseProxyServerTest.java:32]: Please use an ephemeral port to avoid port collision

File: examples/grpc/src/test/java/example/armeria/grpc/GrpcReverseProxyServerTest.java [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+import example.armeria.grpc.Hello.HelloReply;
+import example.armeria.grpc.Hello.HelloRequest;
+
+class GrpcReverseProxyServerTest {
+
+    private static Server server;
+
+    @Container
+    private static GenericContainer<?> envoy = new GenericContainer<>("envoyproxy/envoy:v1.22.0")
+            .withExposedPorts(9090)
```
Comment (trustin) [inline examples/grpc/src/test/java/example/armeria/grpc/GrpcReverseProxyServerTest.java:26]: Please use a random unused port to avoid port collision as much as possible. Use `PortUtil.unusedTcpPort`.

File: examples/grpc/build.gradle (hunk 2 above)
Comment (eottabom) [reply examples/grpc/build.gradle:19]: Well, am I right to understand?
Are you telling me to create and use a grpc-reverse-proxy project in the sample directory?

**Comment by eottabom** [comment]
It does not appear to be a problem caused by overlapping ports.
Let me find the cause again.

**Comment by eottabom** [comment]
build-self-hosted-unsafe-jdk-21-snapshot-blockhound

I don't think there's a docker on the ci server

`java.util.concurrent.CompletionException: java.lang.IllegalStateException: Could not find a valid Docker environment. Please see logs and check configuration`

File: examples/reverse-proxy/src/main/java/example/armeria/reverseproxy/Main.java [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+import com.linecorp.armeria.server.annotation.ProducesJson;
+import com.linecorp.armeria.server.annotation.decorator.LoggingDecorator;
+import com.linecorp.armeria.server.grpc.GrpcService;
+
+public final class Main {
+
+    private static final Logger logger = LoggerFactory.getLogger(Main.class);
+
+    private static final int serverPort = PortUtil.unusedTcpPort();
+    private static final int backendServerPort = PortUtil.unusedTcpPort();
```
Comment (trustin) [inline examples/reverse-proxy/src/main/java/example/armeria/reverseproxy/Main.java:25]: In an example, we should use the fixed ports. I'd expect tests use unused random ports though.

File: examples/reverse-proxy/src/test/java/example/armeria/reverseproxy/GrpcReverseProxyServerTest.java [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+                    .withStartupTimeout(Duration.ofSeconds(60));
+        } catch (IOException ex) {
+            logger.error("Failed to configure Envoy container", ex);
+        }
+    }
+
+    @BeforeAll
+    static void startServer() {
+        server = Server.builder()
+                .http(serverPort)
```
[the remark this answers is not in the document]
Comment (eottabom) [reply examples/reverse-proxy/src/test/java/example/armeria/reverseproxy/GrpcReverseProxyServerTest.java]: I modified it to use port 0.
https://github.com/line/armeria/pull/5722/commits/ce6ab185e1024dc782f7431454c5b96a9df17c05

File: examples/reverse-proxy/src/main/java/example/armeria/reverseproxy/Main.java (hunk 5 above)
Comment (eottabom) [reply examples/reverse-proxy/src/main/java/example/armeria/reverseproxy/Main.java]: I modified it to use a fixed port.

https://github.com/line/armeria/pull/5722/commits/9424ddbc89eddaa53cc7193a6771d9bac5955284

**Comment by jrhee17** [comment]
Pushed some small commits 🙇  Let me know if any changes don't make sense

**Review by jrhee17 — APPROVED** [review]
👍 👍 👍

File: examples/grpc-envoy/src/main/java/example/armeria/grpc/envoy/Main.java [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+    }
+
+    private static Server startBackendServer(int serverPort) {
+        return Server.builder()
+                .http(serverPort)
+                .service(GrpcService.builder()
+                        .addService(new HelloService())
+                        .build())
+                .build();
+    }
```
Comment (minwoox) [inline examples/grpc-envoy/src/main/java/example/armeria/grpc/envoy/Main.java:43]: nit: style
```suggestion
    private static Server startBackendServer(int serverPort) {
        return Server.builder()
                     .http(serverPort)
                     .service(GrpcService.builder()
                                         .addService(new HelloService())
                                         .build())
                     .build();
    }
```

**Review by minwoox — COMMENTED** [review]
Looks great, thanks!
Could you also run this script to see if everything works correctly?
https://github.com/line/armeria-examples/blob/main/update-examples.sh
You might run this command: (I supposed `armeria` and `armeria-examples` are in the same parent directory)
`./update-examples.sh 1.29.0 ../armeria`

File: examples/grpc-envoy/src/main/resources/envoy/envoy.yaml [hunk 8]
```diff
@@ -48,5 +48,5 @@ static_resources:
               - endpoint:
                   address:
                     socket_address:
-                      address: host.testcontainers.internal
+                      address: host.docker.internal
```
Comment (minwoox) [inline examples/grpc-envoy/src/main/resources/envoy/envoy.yaml:51]: `host.docker.internal` doesn't work on Linux. I think we should revert this change.
See https://github.com/envoyproxy/java-control-plane/pull/303

Comment (eottabom) [reply examples/grpc-envoy/src/main/resources/envoy/envoy.yaml]: I was merging because of a crash, but I forgot the file. I fixed it!

**Comment by eottabom** [comment]
> Looks great, thanks! Could you also run this script to see if everything works correctly? https://github.com/line/armeria-examples/blob/main/update-examples.sh You might run this command: (I supposed `armeria` and `armeria-examples` are in the same parent directory) `./update-examples.sh 1.29.0 ../armeria`

This script is not working properly.

The following error occurs
Is there a dependency I need to add?
image

**Comment by minwoox** [comment]
> This script is not working properly.

Let me see what happened. 😉

**Comment by minwoox** [comment]
> Let me see what happened. 😉

It works correctly. The problem was just that we needed to convert the `libs.testcontainers.junit.jupiter` to the actual names in the script like:
`-pe "s/libs.testcontainers.junit.jupiter/'org.testcontainers:junit-jupiter:$TESTCONTAINER_VERSION'/g;" \`

It will be handled when we release the next version so you don't have to worry about it. 😉

**Review by minwoox — APPROVED** [review]
Thanks a lot, @eottabom! 😄

**Comment by eottabom** [comment]
Rather, I think learned a lot. Thank you. 😄

**Review by trustin — APPROVED** [review]
:+1: :bow:

**Review by ikhoon — APPROVED** [review]
Looks a useful example! 🙇‍♂️🙇‍♂️
